### PR TITLE
Fixed Euler 17 and made it more robust and general

### PR DESCRIPTION
**Pull request by paplorinc**
Generalizes https://github.com/javaslang/javaslang/pull/1424.
While the original problem didn't ask for millions/billions, generalizing it wasn't that difficult.
This doesn't showcase the pattern matching anymore, but is still heavily based on Javaslang API.

@jorander, @danieldietrich, could we apply pattern matching internally somewhere?

### CODE REVIEW

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java [hunk 1]
```diff
@@ -5,86 +5,86 @@
...
+     * The use of "and" when writing out numbers is in compliance with British usage.
      */
     @Test
     public void shouldSolveProblem17() {
-        assertThat(letterCountInNumber(1)).isEqualTo(3);
-        assertThat(letterCountInNumber(342)).isEqualTo(23);
-        assertThat(letterCountInNumber(115)).isEqualTo(20);
-        assertThat(letterCountInNumbersUpToAndIncluding(5)).isEqualTo(19);
-        assertThat(letterCountInNumbersUpToAndIncluding(1000)).isEqualTo(21124);
+        List.of("one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", "twenty", "twentyone")
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: list of known values with indices for the first few (problematic) values.

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java [hunk 2]
```diff
@@ -5,86 +5,86 @@
...
-    private static final int THOUSAND = "thousand".length();
-    private static final int HUNDRED = "hundred".length();
-    private static final int AND = "and".length();
-    private static final int[] LETTER_COUNT = new int[91];
+    private static final String CONJUNCTION = "and";
+    private static final Map<Integer, String> LENGTHS = TreeMap.ofEntries(
+            Tuple.of(1, "one"), Tuple.of(2, "two"), Tuple.of(3, "three"), Tuple.of(4, "four"), Tuple.of(5, "five"), Tuple.of(6, "six"), Tuple.of(7, "seven"), Tuple.of(8, "eight"), Tuple.of(9, "nine"),
+            Tuple.of(10, "ten"), Tuple.of(11, "eleven"), Tuple.of(12, "twelve"), Tuple.of(13, "thirteen"), Tuple.of(14, "fourteen"), Tuple.of(15, "fifteen"), Tuple.of(16, "sixteen"), Tuple.of(17, "seventeen"), Tuple.of(18, "eighteen"), Tuple.of(19, "nineteen"),
+            Tuple.of(20, "twenty"), Tuple.of(30, "thirty"), Tuple.of(40, "forty"), Tuple.of(50, "fifty"), Tuple.of(60, "sixty"), Tuple.of(70, "seventy"), Tuple.of(80, "eighty"), Tuple.of(90, "ninety"),
+            Tuple.of(100, "hundred"), Tuple.of(1000, "thousand"), Tuple.of(1_000_000, "million"), Tuple.of(1_000_000_000, "billion")
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: enabled all the values from `Integer`

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java [hunk 3]
```diff
@@ -5,86 +5,86 @@
...
-        return Match(num).of(
-                Case($(n -> n / 1000 > 0), n -> LETTER_COUNT[n / 1000] + THOUSAND + letterCountInNumber(n % 1000)),
-                Case(allOf((Integer n) -> n / 100 > 0, n -> n % 100 > 0), n -> LETTER_COUNT[n / 100] + HUNDRED +  AND + letterCountInNumber(n % 100)),
-                Case($(n -> n / 100 > 0), n -> LETTER_COUNT[n / 100] + HUNDRED),
-                Case($(n -> n / 10 > 1), n -> LETTER_COUNT[n / 10 * 10] + letterCountInNumber(n % 10)),
-                Case($(), n -> LETTER_COUNT[n])
-        );
+    private static String asText(int number) {
+        return LENGTHS.foldRight(Tuple.of(Vector.<String> empty(), number), (magnitudeAndText, lengthsAndRemainder) -> {
+            final int magnitude = magnitudeAndText._1;
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: multiple assignment would really help here :/

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java [hunk 4]
```diff
@@ -5,86 +5,86 @@
...
+                                                   : lengthsAndRemainder;
+        })._1.mkString();
     }
 
-    private static int letterCountInNumbersUpToAndIncluding(int num) {
-        return Stream.rangeClosed(1, num)
-                .map(Euler17Test::letterCountInNumber)
-                .sum().intValue();
+    private static Tuple2<Vector<String>, Integer> asText(int magnitude, String text, Vector<String> chunks, int remainder) {
+        if (needsScalarPrefix(remainder)) {
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: could we pattern match here instead :)?

**Comment by danieldietrich** [comment]
I would like to see @jorander the Pattern Matching showcase here :`/

Idea: Could we make an abstract class containing the optimized data structure and provide two implementations, the generalized one and the one with pattern-matching?

**Comment by paplorinc** [comment]
Not really, but we could use pattern matching somewhere - or here, if it fits.

**Comment by jorander** [comment]
Hi,

I think the two solutions can be combined. I was thinking of using a Map at
first, but thought the Option returned from Map#get was a bit clunky in
this context. But I can wrap that in a function. I'll make a try at
combining the solutions.

/Jörgen
Den 9 jul 2016 20:01 skrev "Daniel Dietrich" <notifications@github.com>:

I would like to see @jorander <https://github.com/jorander> the Pattern
Matching showcase here :`/

Idea: Could we make an abstract class containing the optimized data
structure and provide two implementations, the generalized one and the one
with pattern-matching?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/javaslang/javaslang/pull/1429#issuecomment-231547263>,
or mute the thread
<https://github.com/notifications/unsubscribe/ADRVwFqJnBcivouqxdidjNKGGuXPs5_oks5qT-IRgaJpZM4JIpSo>
.

**Comment by paplorinc** [comment]
> I'll make a try at combining the solutions.

Please leave comments on how I should change this instead :)

**Comment by jorander** [comment]
It's hard to explain the changes needed, but here is a link to what a combination could look like: https://github.com/jorander/javaslang/blob/229e38cdf37788052c19e8cd9f460cbd2bc0bcbd/javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java

**Comment by paplorinc** [comment]
@danieldietrich, opinions? :)

**Comment by danieldietrich** [comment]
@paplorinc Let's take the combined solution (A/B) Jörgen [posted](https://github.com/jorander/javaslang/blob/229e38cdf37788052c19e8cd9f460cbd2bc0bcbd/javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java). I love it :)

**Comment by jorander** [comment]
Glad you liked it. I can make a PR for it if you like.
BR
Jörgen
Den 11 jul 2016 21:56 skrev "Daniel Dietrich" <notifications@github.com>:

**Comment by paplorinc** [comment]
will rather update my PR :)

**Comment by jorander** [comment]
Fine with me.

**Comment by paplorinc** [comment]
Which of the two pattern matches do you guys like more:
```java
return Match(num).of( /*@formatter:off*/
        Case($(n -> n >= 1000),                                n -> length(n / 1000) + length(1000) + letterCount(n % 1000)),
        Case(allOf((Integer n) -> n >= 100, n -> n % 100 > 0), n -> length(n / 100) + length(100) + CONJUNCTION.length() + letterCount(n % 100)),
        Case($(n -> n >= 100),                                 n -> length(n / 100) + length(100)),
        Case($(n -> n >= 20),                                  n -> length(n / 10 * 10) + letterCount(n % 10)),
        Case($(0),                                             0),
        Case($(),                                              n -> length(n))
); /*@formatter:on*/
```
or
```java
return Match(num).of( /*@formatter:off*/
        Case(n -> n >= 1000, n -> length(n / 1000) + length(1000) + letterCount(n % 1000)),
        Case(n -> n >= 100,  n -> Match(n).of(
                                   Case(n1 -> n1 % 100 > 0, n2 -> length(n2 / 100) + length(100) + CONJUNCTION.length() + letterCount(n2 % 100)),
                                   Case($(),                length(n / 100) + length(100)))),
        Case(n -> n >= 20,   n -> length(n / 10 * 10) + letterCount(n % 10)),
        Case(0,              0),
        Case($(),            n -> length(n))
); /*@formatter:on*/
```

i.e. in the second example there's a pattern match *inside* another pattern match, in order to check whether it both greater than 100 and divisible, without duplicating any of the conditions.
Also, thanks to the `@formatter`s both pattern matches will keep this alignment.

@jorander, @danieldietrich, opinions :)?

**Comment by danieldietrich** [comment]
I like the first one more because in the second one the nested `$()` is _indirectly_ determined by the parent match case, which is harder to read.

We can simplify the predicate of the first proposal:

```java
Case(n -> n >= 100 && n % 100 > 0, ...)
```

**Comment by paplorinc** [comment]
> is indirectly determined by the parent 

I consider that a feature, two separate cases refer to the same parent state, i.e. when `n >= 100`.
Otherwise we have to copy paste, which isn't DRY :/

**Comment by danieldietrich** [comment]
We can take the second solution. For me this indentation is a bit better readable (but we also can take yours):

```java
return Match(num).of( /*@formatter:off*/
        Case(n -> n >= 1000,         n -> length(n / 1000) + length(1000) + letterCount(n % 1000)),
        Case(n -> n >= 100,          n -> Match(n).of(
            Case(n1 -> n1 % 100 > 0, n1 -> length(n1 / 100) + length(100) + CONJUNCTION.length() + letterCount(n1 % 100)),
            Case($(),                length(n / 100) + length(100)))),
        Case(n -> n >= 20,           n -> length(n / 10 * 10) + letterCount(n % 10)),
        Case(0,                      0),
        Case($(),                    n -> length(n))
); /*@formatter:on*/
```

**Comment by danieldietrich** [comment]
Btw - that would be a cool feature - having nested `Case`s like this (without typing `n -> Match(n).of(`:

```java
        Case(n -> n >= 100,
            Case(n1 -> n1 % 100 > 0, n1 -> length(n1 / 100) + length(100) + CONJUNCTION.length() + letterCount(n1 % 100)),
            Case($(),                length(n / 100) + length(100)))),
        Case(n -> n >= 20,           n -> length(n / 10 * 10) + letterCount(n % 10)),
```

**Comment by paplorinc** [comment]
Indeed, that's why I experimented with them :)

**Comment by paplorinc** [comment]
Pushed, @jorander, are you okay with this version? :)

File: javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java [hunk 5]
```diff
@@ -30,61 +31,124 @@
...
-        LETTER_COUNT[70] = "seventy".length();
-        LETTER_COUNT[80] = "eighty".length();
-        LETTER_COUNT[90] = "ninety".length();
+    private static void runTestsFor(SolutionProblem17 solution) {
+        List.of("one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", "twenty", "twentyone")
+            .zipWithIndex().forEach(t -> {
+            final int number = t._2.intValue() + 1;
+            final String numberAsString = t._1;
+            assertThat(numberAsString).hasSize(solution.letterCount(number));
+        });
```
Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: the alignment of lines 44-47 should be +4 spaces I think

Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler17Test.java]: This is how Idea formats it, if I put the forEach on that line.
Will put in on the next :)

**Comment by danieldietrich** [comment]
(Offtopic/new feature):

Looking at nested Cases, `Case(..., Case...)` would not make sense because we need at least one case.

`Case(..., Case, Case...)` would be ambiguous with `Case(..., Function)` I think if there is only one nested case.

So nested cases will only work with 2+ nested cases: `Case(..., Case, Case, Case...)` .

(I think so - have to test it)

**Comment by danieldietrich** [comment]
Great, thank you!

**Pull request merged**